### DISCUSSION TITLE
Composition Function

### DISCUSSION THREAD

**Discussion by danishk6100** [crossplane/crossplane#3821] [Q&A]
Hi All,

I am new to crossplane. I wanted to explore a way for editing composition with help of python code. I saw a new feature composition function is available for that but I am not able to understand how to use it. Very less documentation is available for it. 

Could u please help me in how composition function works and what all steps needs to be followed for using it. Also I need  to know how to apply functionIO.

Note: I am having crossplane installed with upbound aws-provider.

**Comment by jbw976** [comment]
Thanks for sharing your interest in the new alpha Composition Functions feature @danishk6100 - did you already get a chance to look at this guide? It contains a fair amount of details about the feature, so you may find it helpful.

https://docs.crossplane.io/knowledge-base/guides/composition-functions/

Let us know any other feedback you have about the functionality and experience of composition functions - we're tracking what is needed to keep maturing it in https://github.com/crossplane/crossplane/issues/3751 🙇

**Comment by danishk6100** [comment]
Hi @jbw976,

I have gone through this documentation. When i am trying to apply the functionIO it is throwing error and its CRD is not present. how can I install its CRD.

Also when I try to run the function by applying claim without applying functionIO it is throwing this error:

Warning  ComposeResources   19m (x246 over 4h39m)    defined/compositeresourcedefinition.apiextensions.crossplane.io  cannot compose resources: cannot run Composition Function pipeline: cannot run function "python-demo": cannot run container: rpc error: code = Unknown desc = exit status 1: xfn: error: spark.Command.Run(): OCI runtime error: exit status 1 Normal   SelectComposition  4m50s (x260 over 4h39m)  defined/compositeresourcedefinition.apiextensions.crossplane.io  Successfully selected composition

In python-demo I am having a simple readIO and writeIO function.

**Comment by jbw976** [comment]
Could you clarify what you mean by "apply the functionIO"? FunctionIO is not actually a CRD that is present as a type in the API server.    
 I realize that can be a little confusing 😊 

From the [XFN Guide](https://docs.crossplane.io/knowledge-base/guides/composition-functions/#functionio):

>  When a Composition Function runner like xfn runs your Function it will write FunctionIO to its stdin. A FunctionIO is a Kubernetes style YAML manifest. It’s not a custom resource (it never gets created in the API server) but it follows Kubernetes conventions.

The error you are seeing is a fairly general one for when something goes wrong with running your function.  Some folks have been discussing their experience with that same error in https://github.com/crossplane/crossplane/issues/3807 - check out that issue to see if it gives you some new insight 🙏 

Some other thoughts:

* have you turned on debug logging for Crossplane and XFN pod/containers?
* here's the command someone tested their function successfully with just Docker: https://github.com/crossplane/crossplane/issues/3818#issuecomment-1463776934
* here's how @negz has been testing the running of his functions using the crossplane/xfn image: https://github.com/crossplane/crossplane/issues/3807#issuecomment-1450905399
